analysis/pseudobulk-bulk-prediction/notebooks/perform-ora.Rmd

```diff
@@ -0,0 +1,320 @@
+      summary_residual = sample(summary_residual_df$summary_residual, nrow(summary_residual_df))   
+    )
+    
+    geneset_list |>
+      purrr::map(
+        \(gs_df) analyze_outliers(shuffled_df, gs_df$ensembl_id)
+      ) |>
+      purrr::list_rbind(names_to = "geneset_cell_type") |>
+      # rename columns to with shuffled_ prefix for code clarity
+      dplyr::rename(
+        shuffled_or_bulk = or_bulk, 
+        shuffled_or_pseudobulk = or_pseudobulk
+      ) 
+  }) |> purrr::list_rbind(names_to = "replicate")
+```
+
+
+Below, we visualize the shuffled odds ratio distributions, with a red line representing the odds ratio calculated from the real data.
+
+```{r fig.height=10, fig.width=10}
+ggplot(shuffled_or_df) + 
+  aes(x = shuffled_or_bulk) + 
+  geom_histogram(binwidth = 0.2, fill = "gray65") +
+  facet_wrap(vars(geneset_cell_type), scales = "free_x") +
+  geom_vline(
+    data = baseline_or_df, 
+    aes(xintercept = or_bulk), 
+    color = "firebrick", 
+    linewidth = 0.75
+  ) + 
+  labs(
+    title = "Odds ratio distributions in the bulk direction", 
+    x = "Shuffled bulk-direction odds ratios"
+  )
+```
+
+
+```{r fig.height=10, fig.width=10}
+ggplot(shuffled_or_df) + 
+  aes(x = shuffled_or_pseudobulk) + 
+  geom_histogram(binwidth = 0.2, fill = "gray65") +
+  facet_wrap(vars(geneset_cell_type), scales = "free_x") +
+  geom_vline(
+    data = baseline_or_df, 
+    aes(xintercept = or_pseudobulk), 
+    color = "firebrick", 
+    linewidth = 0.75
+  ) + 
+  labs(
+    title = "Odds ratio distributions in the pseudobulk direction", 
+    x = "Shuffled pseudobulk-direction odds ratios"
+  )
+```
+
+
+Next, we can calculate P-values for each gene set.
+We'll also adjust P-values for multiple testing using FDR (Benjamini-Hochberg).
+
+```{r}
+permutation_pvalues <- geneset_names |>
+  purrr::set_names(geneset_names) |>
+  purrr::map(
+    \(geneset_name) {
+      
+    # This will have one row for the given geneset
+    geneset_baseline_df <- baseline_or_df |>
+      dplyr::filter(geneset_cell_type == geneset_name)
+    
+    # Define these variables for code clarity
+    baseline_bulk_or <- geneset_baseline_df$or_bulk
+    baseline_pseudobulk_or <- geneset_baseline_df$or_pseudobulk
+      
+    shuffled_or_df |>
+      dplyr::filter(geneset_cell_type == geneset_name) |>
+      # Calculate P-value as fraction of shuffled ORs >= baseline OR
+      dplyr::summarize(
+        bulk = sum(shuffled_or_bulk >= baseline_bulk_or) / params$reps, 
+        pseudobulk = sum(shuffled_or_pseudobulk >= baseline_pseudobulk_or) / params$reps
+      ) |>
+      # Correct for multiple testing with FDR
+      tidyr::pivot_longer(
+        everything(),
+        names_to = "modality", 
+        values_to = "p_value"
+      ) |>
+      dplyr::mutate(
+        # first ensure p-value is at least 1/(params$reps+1)
+        p_value = max(p_value, 1/(params$reps + 1)),
+        p_adj_bh = p.adjust(p_value, method = "BH")
```

Comment:
Ah, here is the error: this would need to be done `rowwise` or you just get the max p. Probably better to just calculate the p value by the more conservative formula of `(n_extreme + 1)/(n_reps +1) from the start.

Reply:
The number of times I have been gotten by needing a `rowwise` is non-zero.

Reply:
or `pmax()` in this case.